[cef-build] Create a new port

serg06 opened the pull request:
**Describe the pull request**

[CEF](https://bitbucket.org/chromiumembedded/cef/src/master/) was requested 5 years ago but nobody has added it yet: #332 

This is the next best thing: [Pre-built CEF binaries](https://cef-builds.spotifycdn.com/index.html).

#### What does your PR fix?  

Fixes #332 (for win64 builds.)

#### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?

x64-windows, Yes

(There are binaries for other triplets on their website, but I didn't implement that.)

#### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  

Yes

#### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  

Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

serg06 committed:
Add cef-build port

serg06 committed:
CI system skip irrelevant builds

serg06 committed:
run format-manifest

serg06 committed:
run x-add-version

serg06 committed:
clean portfile

serg06 committed:
x-add-version

serg06 commented:
Update: I realize I'm not copying the `dll`, `bin`, `pak` and `dat` files correctly. It only ran for me because they were left over from an old build.

How do I get the `pak` and `dat` files to end up in the .exe directory?

JackBoosY commented on ports/cef-build/portfile.cmake:
```diff
@@ -0,0 +1,95 @@
+file(
+	COPY
+		"${RELEASE_BUILD_DIR}/libcef_dll_wrapper/Release/libcef_dll_wrapper.lib"
+		"${RELEASE_BUILD_DIR}/libcef_dll_wrapper/Release/libcef_dll_wrapper.pdb"
+		"${SOURCE_PATH}/Release/libcef.lib"
+	DESTINATION "${CURRENT_PACKAGES_DIR}/lib"
+)
+
+# /bin release
+file(
+	COPY "${SOURCE_PATH}/Release/libcef.dll"
+	DESTINATION "${CURRENT_PACKAGES_DIR}/bin"
+)
+
+# /lib debug
+file(
+	COPY
+		"${DEBUG_BUILD_DIR}/libcef_dll_wrapper/Debug/libcef_dll_wrapper.lib"
+		"${DEBUG_BUILD_DIR}/libcef_dll_wrapper/Debug/libcef_dll_wrapper.pdb"
+		"${SOURCE_PATH}/Debug/libcef.lib"
+	DESTINATION "${CURRENT_PACKAGES_DIR}/debug/lib"
+)
+
+# /bin debug
+file(
+	COPY "${SOURCE_PATH}/Debug/libcef.dll"
+	DESTINATION "${CURRENT_PACKAGES_DIR}/debug/bin"
+)
+
+# /include
+file(
+	COPY "${SOURCE_PATH}/include"
+	DESTINATION "${CURRENT_PACKAGES_DIR}"
+)
+
+# Another /include for cef's own imports
+file(
+	COPY "${SOURCE_PATH}/include"
+	DESTINATION "${CURRENT_PACKAGES_DIR}/include"
+)
```
Why not use `vcpkg_install_cmake` instead? Does the source have `install` code?

serg06 replied:
@JackBoosY It doesn't have `install` code unfortunately, I tried `vcpkg_install_cmake` but it told me "invalid target install".

JackBoosY replied:
Yeah, therefore, we need to patch the source to add the install command.

Pospelove commented:
Please take a look at https://github.com/microsoft/vcpkg/pull/10102

serg06 commented:
@Pospelove But doesn't the recent [chartdir port](https://github.com/microsoft/vcpkg/blob/master/ports/chartdir/portfile.cmake) use prebuilts?

serg06 committed:
install files with INSTALL instead of file(COPY

serg06 committed:
Fix version

serg06 commented:
@JackBoosY I changed it to use `INSTALL()` command for everything, but I still have the same issue, I don't know how to copy `.bin` files to output directory.

For example I added this to CMakeLists.txt:

```
file(GLOB bins_deb Debug/*.dll Debug/*.bin)
file(GLOB bins_rel Release/*.dll Release/*.bin)
INSTALL(
  FILES ${bins_deb}
  CONFIGURATIONS Debug
  TYPE BIN
)
```

And they appear in the bin folder:

![image](https://user-images.githubusercontent.com/1350889/123567843-1192df80-d791-11eb-9cb6-330dbd582dcf.png)

But they don't get copied over to .exe directory. Am I doing something wrong?

JackBoosY commented:
@serg06  I will give you some help.

JonLiu1993 commented:
@serg06 ,Can you please resolve the conflicts against master?

JackBoosY commented:
I noticed that this port only generates the library `libcef_dll_wrapper`, and other libraries copy existing files. Right?
If it's true, we hope that the main library `cef` is generated by the build process and change the port name to cef.

serg06 commented:
@JackBoosY Yeah, this library copies a bunch of pre-built binaries and generates `libcef_dll_wrapper`.

Are you suggesting we generate the `cef` binaries ourselves instead of using prebuilts? How could that help with the current issue?

JackBoosY commented:
@serg06 According to our policy, we need to generate binaries as much as possible instead of directly installing pre-built binaries.

serg06 commented:
@JackBoosY That's fine, but it has been years and nobody has added cef yet. Wouldn't it better to have prebuilt binaries than to have nothing at all?

JackBoosY commented:
@serg06 Sorry but according to our policy, only SDK can use pre-built binaries.
We should use vcpkg_configure_gn to build it.
Example: [chromium-base](https://github.com/microsoft/vcpkg/blob/master/ports/chromium-base/portfile.cmake)

serg06 commented:
@JackBoosY Ah okay. Thanks for the help. I hope someone will add it in the future.
